idaes/apps/grid_integration/tracker.py

```diff
@@ -0,0 +1,1951 @@
+
+        Arguments:
+            tracking_model_object: the initialized model object for tracking
+            n_tracking_hour: number of implemented hours after each solve
+            solver: a Pyomo mathematical programming solver object
+
+        Returns:
+            None
+        '''
+
+        self.tracking_model_object = tracking_model_object
+        self.n_tracking_hour = n_tracking_hour
+        self.solver = solver
+        self.model = self.tracking_model_object.model
+        self.formulate_tracking_problem()
+
+    def formulate_tracking_problem(self):
+
+        '''
+        Formulate the tracking optimization problem by adding necessary
+        parameters, constraints, and objective function.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        '''
+
+        # get the sets needed for the tracking problem
+        self.tracking_set = self._get_tracking_sets()
+
+        self._add_tracking_params()
+        self._add_tracking_constraints()
+        self._add_tracking_objective()
+
+        return
+
+    def _get_tracking_sets(self):
+
+        '''
+        Get the necessary index sets for tracking, i.e., generator names and
+        time.
+
+        Arguments:
+            None
+
+        Returns:
+            tracking_set: a list that contains the necessary Pyomo set objects.
+        '''
+
+        tracking_set = []
+
+        model_sets = self.tracking_model_object.indices
+
+        for s, n in model_sets.items():
+            if n in self.tracking_set_names:
+                tracking_set.append(s)
+
+        return tracking_set
+
+    def _add_tracking_params(self):
+
+        '''
+        Add necessary tracking parameters to the model, i.e., market dispatch
+        signal.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        '''
+
+        # add params to the model
+        self.model.power_dispatch = pyo.Param(*self.tracking_set, \
+                                              initialize = 0, \
+                                              mutable = True)
+        return
+
+    def _add_tracking_constraints(self):
+
+        '''
+        Add necessary tracking constraints to the model, e.g., power output needs
+        to follow market dispatch signals.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        '''
+
+        self._add_tracking_dispatch_constraints()
+        return
+
+    def _add_tracking_dispatch_constraints(self):
+
+        '''
+        Add tracking constraints to the model, i.e., power output needs
+        to follow market dispatch signals.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        '''
+
+        # declare a constraint list
+        self.model.tracking_dispatch_constraints = pyo.ConstraintList()
+
+        # unpack things
+        model_sets = self.tracking_model_object.indices
+        set_ls = list(model_sets.keys())
```

Comment:
Does it make sense to define an `unpack` method in each class. Looks like this is used several places.